File: go/euler50.go
package main
import "fmt"

func sieve(n int) []int {
    primes := []int{}
    sieved := make([]bool, n)
    for i := 2; i < n; i++ {
        if !sieved[i] {
            primes = append(primes, i)
            for j := 2 * i; j < n; j += i {
                sieved[j] = true
            }
        }
    }
    return primes
}

func main() {
    res := 0
    primes := sieve(1000000)
    primesSet := make(map[int]bool)
    sums := []int{0}
    maxSum := 0
    for _, p := range primes {
        if maxSum + p < 1000000 {
            maxSum += p
            sums = append(sums, maxSum)
        }
        primesSet[p] = true
    }
    
    // use maximum window size, shrink until prime is found with current window size
    for length := len(sums); res == 0 && length > 0; length-- {
        // use latest start index to guarantee first prime sum found is largest
        // start + length < len(sums)
        // start + length <= len(sums) - 1
        // 0 <= start <= len(sums) - 1 - length
        for start := len(sums) - 1 - length; res == 0 && start >= 0; start-- {
            sumA := sums[start]
            sumB := sums[start + length]
            total := sumB - sumA
            if primesSet[total] {
                res = total
            }
        }
    }
    fmt.Println(res)
}

File: ts/euler50.ts
function sieve(): number[] {
    const primes: number[] = [];
    const sieved = Array(1_000_000).fill(false);
    for (let i = 2; i < sieved.length; i++) {
        if (!sieved[i]) {
            primes.push(i);
            for (let j = i * i; j < sieved.length; j += i) {
                sieved[j] = true;
            }
        }
    }
    return primes;
}

function consecutivePrimeSum(): number {
    const primes = sieve();
    const primesSet = new Set(primes);
    const prefixSums = [0];
    let curr = 0;
    for (let prime of primes) {
        curr += prime;
        if (curr > 1_000_000) break;
        prefixSums.push(curr);
    }
    const n = prefixSums.length;
    // number of primes in sum
    for (let length = prefixSums.length - 1; length >= 0; length--) {
        // last index of prime counted in sum
        for (let end = n - 1; end - length >= 0; end--){
            // first prime sum found is guaranteed to be largest
            const sum = prefixSums[end] - prefixSums[end - length];
            if (primesSet.has(sum)) return sum;
        }
    }
    return -1;
}

console.log(consecutivePrimeSum());
